crates/transport/src/behaviour/stream_client.rs

```diff
@@ -0,0 +1,238 @@
+
+#[derive(Debug, Clone, Copy)]
+pub struct ClientConfig {
+    /// The maximum number of open substreams per peer (default: 3)
+    pub max_concurrent_streams: usize,
+    /// Timeout applied on dialing the remote and opening a substream (default: 10 sec)
+    pub connect_timeout: Duration,
+    /// Timeout applied on the entire request-response cycle (default: 60 sec)
+    pub request_timeout: Duration,
+}
+
+impl Default for ClientConfig {
+    fn default() -> Self {
+        Self {
+            max_concurrent_streams: 3,
+            connect_timeout: Duration::from_secs(10),
+            request_timeout: Duration::from_secs(60),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RequestError {
+    #[error(transparent)]
+    Timeout(Timeout),
+    #[error("the remote supports none of the requested protocols")]
+    UnsupportedProtocol,
+    #[error("response too large")]
+    ResponseTooLarge,
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
+pub enum Timeout {
+    /// Peer lookup or connection establishing timed out
+    #[error("connect timeout")]
+    Connect,
+    /// The response has not been received in time
+    #[error("request timeout")]
+    Request,
+}
+
+pub struct StreamClientHandle {
+    protocol: &'static str,
+    config: ClientConfig,
+    control: libp2p_stream::Control,
+    semaphores: parking_lot::Mutex<HashMap<PeerId, Arc<tokio::sync::Semaphore>>>,
+}
+
+impl StreamClientHandle {
+    pub async fn get_raw_stream(
+        &self,
+        peer: PeerId,
+    ) -> Result<impl AsyncRead + AsyncWrite, RequestError> {
+        let fut = async {
+            let permit = self
+                .semaphores
+                .lock()
+                .entry(peer)
+                .or_insert_with(|| {
+                    Arc::new(tokio::sync::Semaphore::new(self.config.max_concurrent_streams))
+                })
+                .clone()
+                .acquire_owned()
+                .await
+                .expect("Semaphone should never be closed");
+
+            log::debug!("Opening stream to {}", peer);
+            let stream = self
+                .control
+                .clone()
+                .open_stream(peer, StreamProtocol::new(&self.protocol))
+                .await?;
+            log::debug!("Opened stream to {}", peer);
+
+            Ok(StreamWithPayload::new(stream, permit))
+        };
+        tokio::time::timeout(self.config.connect_timeout, fut)
+            .await
+            .unwrap_or_else(|_| {
+                log::debug!("Connection to {peer} timed out");
+                Err(RequestError::Timeout(Timeout::Connect))
+            })
+    }
+
+    /// Send request and get a response in a way compatible with the request-server behaviour
+    pub async fn request_response(
+        &self,
+        peer: PeerId,
+        request: &[u8],
+        max_response_size: u64,
```

Comment:
If a particular instance of this struct is tied to a specific protocol, why pass `max_response_size` with every call, when it can be a part of the config?

Reply:
You're right, I was not sure about the interface here. My thought was that if someone wants to only use the `get_raw_stream` method he shouldn't care about the value of `max_response_size`, that's why I'm only passing it to the `request_response` method. On the other hand, maybe it's not too bad to have an unused value in the config for the sake of convenience in case with calling `request_response`. So agreed, I should move it to the config